pkg/querier/frontend/config.go

```diff
@@ -0,0 +1,111 @@
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	httpgrpc_server "github.com/weaveworks/common/httpgrpc/server"
+
+	"github.com/cortexproject/cortex/pkg/querier"
+	"github.com/cortexproject/cortex/pkg/querier/frontend2"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// This struct combines several configuration options together to preserve backwards compatibility.
+type CombinedFrontendConfig struct {
+	Handler   HandlerConfig    `yaml:",inline"`
+	Frontend  Config           `yaml:",inline"`
+	Frontend2 frontend2.Config `yaml:",inline"`
+
+	CompressResponses bool   `yaml:"compress_responses"`
+	DownstreamURL     string `yaml:"downstream_url"`
+}
+
+func (cfg *CombinedFrontendConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.Handler.RegisterFlags(f)
+	cfg.Frontend.RegisterFlags(f)
+	cfg.Frontend2.RegisterFlags(f)
+
+	f.BoolVar(&cfg.CompressResponses, "querier.compress-http-responses", false, "Compress HTTP responses.")
+	f.StringVar(&cfg.DownstreamURL, "frontend.downstream-url", "", "URL of downstream Prometheus.")
+}
+
+func (cfg *CombinedFrontendConfig) InitFrontend(limits Limits, grpcListenPort int, log log.Logger, reg prometheus.Registerer) (http.RoundTripper, *Frontend, *frontend2.Frontend2, error) {
+	switch {
+	case cfg.DownstreamURL != "":
+		// If the user has specified a downstream Prometheus, then we should use that.
+		rt, err := NewDownstreamRoundTripper(cfg.DownstreamURL)
+		return rt, nil, nil, err
+
+	case cfg.Frontend2.SchedulerAddr != "":
+		// If query-scheduler address is configured, use Frontend2.
+		if cfg.Frontend2.Addr == "" {
+			addr, err := util.GetFirstAddressOf(cfg.Frontend2.InfNames)
+			if err != nil {
+				return nil, nil, nil, errors.Wrap(err, "failed to get frontend address")
+			}
+
+			cfg.Frontend2.Addr = addr
+		}
+
+		if cfg.Frontend2.Port == 0 {
+			cfg.Frontend2.Port = grpcListenPort
+		}
+
+		fr, err := frontend2.NewFrontend2(cfg.Frontend2, log, reg)
+		return AdaptGrpcRoundTripperToHTTPRoundTripper(fr), nil, fr, err
+
+	default:
+		// No scheduler = use original frontend.
+		fr, err := New(cfg.Frontend, limits, log, reg)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		return AdaptGrpcRoundTripperToHTTPRoundTripper(fr), fr, nil, err
+	}
+}
+
+// Configuration for both querier workers, V1 (using frontend) and V2 (using scheduler). Since many flags are reused
+// between the two, they are exposed to YAML/CLI in V1 version (WorkerConfig), and copied to V2 in the init method.
+type CombinedWorkerConfig struct {
+	WorkerV1 WorkerConfig                   `yaml:",inline"`
+	WorkerV2 frontend2.QuerierWorkersConfig `yaml:",inline"`
+}
+
+func (cfg *CombinedWorkerConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.WorkerV1.RegisterFlags(f)
+	cfg.WorkerV2.RegisterFlags(f)
+}
+
+func (cfg *CombinedWorkerConfig) InitQuerierWorker(querierCfg querier.Config, handler http.Handler, log log.Logger) (services.Service, error) {
```

Comment:
I've removed this method from `*CombinedWorkerConfig` (and similarly from `*CombinedFrontendConfig`), but I'd prefer to keep the name, since `New...` indicates creation of new instance or error, but these functions can return `nil` result in some cases, which to me goes against `New...` naming. I also consider them to be extension of initialization done by init methods in `modules.go`.WDYT?

Reply:
LGTM, thanks!